### PR TITLE
Add Domain algebra

### PR DESCRIPTION
**Pull request by dbarbier**
* New classes: DomainComplement, DomainIntersection, DomainUnion, DomainDisjunctiveUnion, DomainDifference. With these new classes, it is easy to build complex domains.
* Add Domain::contains on a Sample,  this may allow some optimizations.
* Deprecate Domain::getLowerBound, getUpperBound.
* Add an argument to KarhunenLoeveQuadratureAlgorithm constructor to pass domain bounds.

### CODE REVIEW

File: python/src/DomainDisjunctiveUnion_doc.i.in [hunk 1]
```diff
@@ -0,0 +1,45 @@
+%feature("docstring") OT::DomainDisjunctiveUnion
+"A Domain representing the disjunctive union of two Domain.
```
Comment (jschueller) [inline python/src/DomainDisjunctiveUnion_doc.i.in]: skip a line here

Comment (dbarbier) [reply python/src/DomainDisjunctiveUnion_doc.i.in]: Thanks, fixed.

**Comment by dbarbier** [comment]
Sorry, one test was failing, and I also discovered that I redefined Mesh::contains(Sample), this change was unintentional.  Commits are squashed.

**Comment by dbarbier** [comment]
Err, this is crazy, I still failed to drop Mesh::contains(Sample) from previous commit, sorry for the trouble.

**Comment by regislebrun** [comment]
@dbarbier Good job. I as wondering if it would be of any interest to make DomainImplementation::contains() parallel? What is your opinion, for this case or for other similar ones:
  + the top class provides a generic parallel implementation of the method and the specialized classes overload it with another implementation if they have trouble with this implementation or if they can use more specific optimizations
  + the top class provides only a sequential implementation and the other classes have to overload it with something better if possible
  + the top class provides three methods: a parallel implementation with the Parallel suffix, a sequential implementation with the Sequential suffix, and a generic implementation pointing to the sequential version. This way, each specialization can choose to use which version to use.

Otherwise, +1

**Comment by dbarbier** [comment]
@regislebrun IMO this is useless:
* Interval::contains code is fast
* LevelSet::contains will benefit from parallel evaluation of functions
* Mesh::contains is dramatically slow, but multithreading won't help, this data structure is not designed for such queries

**Comment by jschueller** [comment]
The tests fail on the autobuilder, probably due to the old python formatting issues

**Comment by dbarbier** [comment]
Indeed, I just added `ot.PlatformInfo.SetNumericalPrecision(6)` as in t_Mesh_std.py; if it still  fails, I will rewrite these tests.

**Comment by regislebrun** [comment]
@dbarbier This code:
```
/* Check if the given point is in the mesh */
Bool Mesh::contains(const Point & point) const
{
  // First, check against the bounding box
  if (!Interval(getLowerBound(), getUpperBound()).contains(point)) return false;
  // Second, check the simplices containing the nearest vertex
  const UnsignedInteger nearestIndex = getNearestVertexIndex(point);
  Indices result(1, nearestIndex);
  // To be sure that the vertices to simplices map is up to date
  if (verticesToSimplices_.getSize() == 0) (void) getVerticesToSimplicesMap();
  const Indices simplicesCandidates(verticesToSimplices_[nearestIndex]);
  for (UnsignedInteger i = 0; i < simplicesCandidates.getSize(); ++i)
    if (checkPointInSimplex(point, simplicesCandidates[i])) return true;
  // Third, a full loop to deal with points not inside of a simplex associated to the nearest vertex 
  const UnsignedInteger simplicesSize = getSimplicesNumber();
  for (UnsignedInteger i = 0; i < simplicesSize; ++i) if (checkPointInSimplex(point, i)) return true;
  return false;
}

/* Check if the given point is in the mesh */
Mesh::BoolCollection Mesh::contains(const Sample & sample) const
{
  const UnsignedInteger size = sample.getSize();
  BoolCollection result(size);
  for (UnsignedInteger i = 0; i < size; ++i)
    result[i] = contains(sample[i]);
  return result;
}
```
provides a significant speed-up on the benchmark given at the end of the comment.=
It can be improved eg by computing the bounding box once and for all. One could also check the second nearest simplices (ie those associated to the vertices of the simplices adjacent to the nearest vertex) but I am not sure it would improve things a lot. On the following benchmark:
```
from openturns import *
from time import time

N = 50
size = 1000

mesh = IntervalMesher([N]*2).build(Interval([0.0]*2, [10.0]*2))
f = SymbolicFunction(["x", "y"], ["x + 0.5*sin(y)", "y-0.1*x*sin(x)"])
mesh.setVertices(f(mesh.getVertices()))

sample = ComposedDistribution([Uniform(-2.0, 12.0)]*2).getSample(size)

t0 = time()
res = mesh.contains(sample)
t1 = time()
print("res=", sum(res), "t=", t1 - t0, "s")
```
less than 0.3% of the points contained by the mesh have been missed by the loop over the candidate. The speed-up is 5x (3.97s vs 19.73s).

**Comment by dbarbier** [comment]
@regislebrun This is still way too slow, compare with `mesh.getNearestVertexIndex(sample)`, and do not forget to add `mesh.computeKDTree()` first.
I did not try your code yet, but am pretty sure that this is caused by missed points. The only solution is to provide another tree specially crafted for cell location.

**Comment by regislebrun** [comment]
and if I use a parallel loop like this:
```
/* TBB policy to speed-up contains over a sample */
struct MeshContainsPolicy
{
  const Sample & sample_;
  Mesh::BoolCollection & result_;
  const Mesh & mesh_;

  MeshContainsPolicy(const Sample & sample,
		     Mesh::BoolCollection & result,
		     const Mesh & mesh)
    : sample_(sample)
    , result_(result)
    , mesh_(mesh)
  {}

  inline void operator()( const TBB::BlockedRange<UnsignedInteger> & r ) const
  {
    for (UnsignedInteger i = r.begin(); i != r.end(); ++i) result_[i] = mesh_.contains(sample_[i]);
  }

}; /* end struct MeshNearestPolicy */

/* Check if the given point is in the mesh */
Mesh::BoolCollection Mesh::contains(const Sample & sample) const
{
  if (sample.getDimension() != getDimension()) throw InvalidArgumentException(HERE) << "Error: expected points of dimension " << getDimension() << ", got points of dimension " << sample.getDimension();
  const UnsignedInteger size = sample.getSize();
  BoolCollection result(size);
  if (size == 0) return result;
  // To be sure that the vertices to simplices map is up to date and
  // not recomputed within the parallel loop
  if (verticesToSimplices_.getSize() == 0) (void) getVerticesToSimplicesMap();
  const MeshContainsPolicy policy(sample, result, *this );
  TBB::ParallelFor( 0, size, policy );
  return result;
}
```
the speed-up is now 23.3x (0.85s vs 19.73s).

**Comment by dbarbier** [comment]
@regislebrun As I said, compare with getNearestVertexIndex, timings should be similar.  And increase N too.

**Comment by regislebrun** [comment]
@dbarbier I never said it was the definite implementation of the method. It is just fast slow code compared to slow code ;-). Of course, a dedicated structure would be a killer, but if we have no time to code it right now, maybe we could use this code in the mean time... It is up to you, no offense.

**Comment by regislebrun** [comment]
@dbarbier I checked [https://en.wikipedia.org/wiki/Point_location](url), and my opinion is to have a specific implementations for dimension 1, 2 and 3 (the most common cases) based on efficient algorithms, and the fast slow code for the higher dimensions as the preprocessing step of accelerated methods looks like a memory killer.

**Comment by dbarbier** [comment]
@regislebrun Your code looks good.  You can also add
```
  const UnsignedInteger dimension = getDimension();
  const Indices vertexIndices(simplices_[index]);
  for (UnsignedInteger i = 0; i < dimension; ++i)
  {
    UnsignedInteger lower = 0;
    UnsignedInteger greater = 0;
    for (UnsignedInteger j = 0; j <= dimension; ++j)
    {
      if(point[i] < vertices_(vertexIndices[j], i))
        ++lower;
      if(point[i] > vertices_(vertexIndices[j], i))
        ++greater;
    }
    if (lower == dimension + 1 || greater == dimension + 1)
      return false;
  }
```
at the beginning of checkPointInSimplexWithCoordinates to exit early if the point is outside the bounding box of the current simplex.

Anyway these improvements can be put in another PR.

**Comment by regislebrun** [comment]
ok. So let's integrate your PR!

**Comment by dbarbier** [comment]
@jschueller It seems that autobuilder did not try to build the last commit

**Comment by jschueller** [comment]
Yes, it was dequeued on finish, restarting manually.

**Comment by dbarbier** [comment]
Still failing, trying other values

**Comment by dbarbier** [comment]
@jschueller Can you please restart it again?

**Comment by jschueller** [comment]
Done

**Comment by dbarbier** [comment]
@regislebrun See BoundingVolumeHierarchy class in my db/BVH branch, it looks promising.

**Comment by regislebrun** [comment]
@dbarbier I saw it, I am one of your attentive followers ;-)! Unfortunately, even if it looks very promising I have no time to check it right now. Hopefully the holidays are coming soon.

**Pull request merged**